keras_cv/losses/giou_loss.py

```diff
@@ -0,0 +1,156 @@
+                f"boxes to be unbatched.  Received `len(boxes1.shape)`={boxes1_rank}, "
+                f"`len(boxes2.shape)`={boxes2_rank}.  Expected either `len(boxes1.shape)`=2 AND "
+                "`len(boxes2.shape)`=2, or `len(boxes1.shape)`=3 AND `len(boxes2.shape)`=3."
+            )
+
+        target = bounding_box.preserve_rel(
+            target_bounding_box_format="yxyx",
+            bounding_box_format=self.bounding_box_format,
+        )
+
+        boxes1 = bounding_box.convert_format(
+            boxes1, source=self.bounding_box_format, target=target
+        )
+
+        boxes2 = bounding_box.convert_format(
+            boxes2, source=self.bounding_box_format, target=target
+        )
+
+        def compute_giou_for_batch(boxes):
+            boxes1, boxes2 = boxes
+            zero = tf.convert_to_tensor(0.0, boxes1.dtype)
+            boxes1_ymin, boxes1_xmin, boxes1_ymax, boxes1_xmax = tf.unstack(
+                boxes1[..., :4], 4, axis=-1
+            )
+            boxes2_ymin, boxes2_xmin, boxes2_ymax, boxes2_xmax = tf.unstack(
+                boxes2[..., :4], 4, axis=-1
+            )
+            boxes1_width = tf.maximum(zero, boxes1_xmax - boxes1_xmin)
+            boxes1_height = tf.maximum(zero, boxes1_ymax - boxes1_ymin)
+            boxes2_width = tf.maximum(zero, boxes2_xmax - boxes2_xmin)
+            boxes2_height = tf.maximum(zero, boxes2_ymax - boxes2_ymin)
+            boxes1_area = boxes1_width * boxes1_height
+            boxes2_area = boxes2_width * boxes2_height
+            intersect_ymin = tf.maximum(boxes1_ymin, boxes2_ymin)
+            intersect_xmin = tf.maximum(boxes1_xmin, boxes2_xmin)
+            intersect_ymax = tf.minimum(boxes1_ymax, boxes2_ymax)
+            intersect_xmax = tf.minimum(boxes1_xmax, boxes2_xmax)
+            intersect_width = tf.maximum(zero, intersect_xmax - intersect_xmin)
+            intersect_height = tf.maximum(zero, intersect_ymax - intersect_ymin)
+            intersect_area = intersect_width * intersect_height
+
+            union_area = boxes1_area + boxes2_area - intersect_area
+            iou = tf.math.divide_no_nan(intersect_area, union_area)
+
+            # giou calculation
+            enclose_ymin = tf.minimum(boxes1_ymin, boxes2_ymin)
+            enclose_xmin = tf.minimum(boxes1_xmin, boxes2_xmin)
+            enclose_ymax = tf.maximum(boxes1_ymax, boxes2_ymax)
+            enclose_xmax = tf.maximum(boxes1_xmax, boxes2_xmax)
+            enclose_width = tf.maximum(zero, enclose_xmax - enclose_xmin)
+            enclose_height = tf.maximum(zero, enclose_ymax - enclose_ymin)
+            enclose_area = enclose_width * enclose_height
+            return iou - tf.math.divide_no_nan(
+                (enclose_area - union_area), enclose_area
+            )
+
+        if boxes1_rank == 2:
+            return compute_giou_for_batch((boxes1, boxes2))
+        else:
+            return tf.vectorized_map(compute_giou_for_batch, elems=(boxes1, boxes2))
```

Comment:
Does vectorized_map work in TPU?

Reply:
Not sure about this. I can't seem to find any particular note in the documentation. 
@LukeWood Can you help here? I will update the code to `map_fn` if this is an issue.

Reply:
I think we should always prefer batched implementation and not rely on vectorized_map or map_fn in loss computation. However sometimes that comes with more work, If your YOLOX implementation does not require > rank 2 inputs, I suggest you can error it out for now